### ISSUE TITLE
[NATS transport] Support for headers in EncodeJSONResponse

### ISSUE THREAD

**Issue by mcosta74** [go-kit/kit#1238]
### What would you like?

# Overview
Since release `v1.11.0`, nats.go supports message headers. headers are very useful to transfer metadata together with messages.

I'm proposing to add basic support for headers in the default response encoder.

# Proposed API

The idea is to do something similar it's done for `transport/http` with the `Headerer` interface. API change proposal is backward compatible (no public API change)

```go
type Headerer interface {
    Headers() Header
}

func EncodeJSONResponse(_ context.Context, reply string, nc *nats.Conn, response interface{}) error {
        msg := nats.NewMessage(reply)
        if nc. HeadersSupported() {
                if headerer, ok := response.(Headerer); ok {
                         msg.Header = headerer.Headers() {
                }
        }

	msg.Data, err := json.Marshal(response)
	if err != nil {
		return err
	}

	return nc.PublishMsg(msg)
}
``` 

If the proposal is accepted I can raise a PR

**Comment by peterbourgon** [comment]
EncodeJSONResponse is a default implementation of a function that consumers can also provide themselves. Does anything prevent that, here?

**Comment by mcosta74** [comment]
No, I'm wondering if a default implementation could have the same feature
than the HTTP one.

==============

Massimo Costa

Il mer 29 giu 2022, 18:51 Peter Bourgon ***@***.***> ha
scritto:

**Comment by peterbourgon** [comment]
So I'd like to try to avoid the situation where default implementations provided by Go kit, like this function, being continuously updated with new features/capabilities of their underlying systems. I see that as unsustainable, as well as unnecessary: Go kit is designed specifically to allow consumers to provide their own EncodeJSONResponse functions.

Can you suggest some wording for the docs that would make this clear?

**Comment by mcosta74** [comment]
Something like

> Go kit provides default implementation for many functions (for instance `http.EncodeJSONResponse`, `nats.EncodeJSONResponse`, `http.DefaultErrorEncoder`, `nats.DefaultErrorEncoder`,...) but is designed to allow consumers to provide their implementation (maybe to use the latest features of the underlying systems). 

?

**Comment by peterbourgon** [comment]
SGTM

**Comment by mcosta74** [comment]
Can you suggest "where" to put this sentence in the documentation?

- website
- README file
- ...
?
